### PR TITLE
Add debug mode and golang profiling

### PR DESCRIPTION
**Pull request by guangbochen**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

Need to add debug mode to the harvester server and enable golang profiling when debug is enabled.

**Solution:**

1. Add debug mode to the Harvester app flag
2. Add golang profiling

**Related Issue:**
https://github.com/rancher/harvester/issues/436

**Test plan:**

N/A

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -69,49 +92,84 @@ func main() {
...
+			Destination: &profileAddress,
+		},
+	}
+	app.Action = func(c *cli.Context) error {
+		// enable profiler
+		if profileAddress != "" {
+			go func() {
+				log.Println(http.ListenAndServe(profileAddress, nil))
+			}()
+		}
```
Comment (gitlawr) [inline main.go:131]: Looks like the profiler is always enabled. But in the chart, it is controlled by the debug flag.

File: pkg/api/keypair/formatter.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ func CollectionFormatter(request *types.APIRequest, collection *types.GenericCol
 type KeyGenActionHandler struct {
 	KeyPairs     v1alpha1.KeyPairClient
 	KeyPairCache v1alpha1.KeyPairCache
+	Config       config.Options
```
Comment (gitlawr) [inline pkg/api/keypair/formatter.go]: We are passing the config everywhere mostly for the namespace. Is it better to store it as a setting?

File: pkg/controller/crds/setup.go [hunk 3]
```diff
@@ -3,6 +3,8 @@ package crds
 import (
 	"context"
 
+	"github.com/rancher/harvester/pkg/config"
```
Comment (gitlawr) [inline pkg/controller/crds/setup.go]: format imports

File: main.go (hunk 1 above)
Comment (guangbochen) [reply main.go:131]: removed chart condition check of profiling.

File: pkg/api/keypair/formatter.go (hunk 2 above)
Comment (guangbochen) [reply pkg/api/keypair/formatter.go]: updated to passing namespace only where it is appliable.

File: pkg/controller/master/image/controller.go [hunk 4]
```diff
@@ -42,6 +43,7 @@ func RegisterController(ctx context.Context, management *config.Management) {
 type Handler struct {
 	images     v1alpha1.VirtualMachineImageClient
 	imageCache v1alpha1.VirtualMachineImageCache
+	cfg        config.Options
```
Comment (gitlawr) [inline pkg/controller/master/image/controller.go]: An abbreviation is okay for a variable but is confusing as the field name

File: pkg/config/context.go [hunk 5]
```diff
@@ -24,17 +24,19 @@ type (
 	_scaledKey struct{}
 )
 
-var (
+type Options struct {
```
Comment (gitlawr) [inline pkg/config/context.go:27]: Is it better to call it ServerOptions?

File: pkg/controller/crds/setup.go [hunk 6]
```diff
@@ -8,10 +8,11 @@ import (
 	wcrd "github.com/rancher/wrangler/pkg/crd"
 
 	"github.com/rancher/harvester/pkg/apis/harvester.cattle.io/v1alpha1"
+	"github.com/rancher/harvester/pkg/config"
 	"github.com/rancher/harvester/pkg/util/crd"
 )
 
-func Setup(ctx context.Context, server *server.Server, controllers *server.Controllers) error {
+func Setup(ctx context.Context, server *server.Server, controllers *server.Controllers, cfg config.Options) error {
```
Comment (gitlawr) [inline pkg/controller/crds/setup.go]: Nit: `s/cfg/opts/` since the struct type is Options

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**